app-service-template/main.go

```diff
@@ -31,45 +34,86 @@ const (
-func CreateAndRunService(serviceKey string, newServiceFactory func(string) (interfaces.ApplicationService, bool)) int {
-	service, ok := newServiceFactory(serviceKey)
+// CreateAndRunAppService wraps what would normally be in main() so that it can be unit tested
+// TODO: Remove and just use regular main() if unit tests of main logic not needed.
+func (app *myApp) CreateAndRunAppService(serviceKey string, newServiceFactory func(string) (interfaces.ApplicationService, bool)) int {
+	var ok bool
+	app.service, ok = newServiceFactory(serviceKey)
 	if !ok {
 		return -1
 	}
 
-	lc := service.LoggingClient()
+	app.lc = app.service.LoggingClient()
+
+	// TODO: Replace with retrieving your custom ApplicationSettings from configuration or
+	//       remove if not using AppSetting configuration section.
+	// For more details see https://docs.edgexfoundry.org/2.0/microservices/application/GeneralAppServiceConfig/#application-settings
+	deviceNames, err := app.service.GetAppSettingStrings("DeviceNames")
+	if err != nil {
+		app.lc.Errorf("failed to retrieve DeviceNames from configuration: %s", err.Error())
+		return -1
+	}
+
+	// More advance custom structured configuration can be defined and loaded as in this example.
+	// For more details see https://docs.edgexfoundry.org/2.0/microservices/application/GeneralAppServiceConfig/#custom-configuration
+	// TODO: Change to use your service's custom configuration struct
+	//       or remove if not using custom configuration capability
+	app.serviceConfig = &config.ServiceConfig{}
+	if err := app.service.LoadCustomConfig(app.serviceConfig, "AppCustom"); err != nil {
+		app.lc.Errorf("failed load custom configuration: %s", err.Error())
+		return -1
+	}
+
+	// Optionally validate the custom configuration after it is loaded.
+	// TODO: remove if you don't have custom configuration or don't need to validate it
+	if err := app.serviceConfig.AppCustom.Validate(); err != nil {
+		app.lc.Errorf("custom configuration failed validation: %s", err.Error())
+		return -1
+	}
 
-	// TODO: Replace with retrieving your custom ApplicationSettings from configuration
-	deviceNames, err := service.GetAppSettingStrings("DeviceNames")
+	// Custom configuration can be 'writable' or a section of the configuration can be 'writable' when using
+	// the Configuration Provider, aka Consul.
+	// For more details see https://docs.edgexfoundry.org/2.0/microservices/application/GeneralAppServiceConfig/#writable-custom-configuration
+	// TODO: Remove if not using writable custom configuration
+	configChanged := make(chan bool)
+	err = app.service.ListenForCustomConfigChanges(&app.serviceConfig.AppCustom, "AppCustom", configChanged)
 	if err != nil {
-		lc.Errorf("failed to retrieve DeviceNames from configuration: %s", err.Error())
+		app.lc.Errorf("unable to watch custom writable configuration: %s", err.Error())
 		return -1
 	}
+	app.serviceConfig.AppCustom.WaitForCustomConfigChanges(configChanged, app.lc)
```

Comment:
Q: the reason for unbuffered channel `configChanged` is because both `ListenForCustomConfigChanges` and `WaitForCustomConfigChanges` are in go routines?  otherwise, probably use buffered channel

Reply:
Yup, no need for buffered